### PR TITLE
Add m2m100

### LINKED ISSUE #8054: Add m2m 100 multilingual translation model from FAIR
Weights, code are available.

+ Fairseq Code: https://github.com/pytorch/fairseq/tree/master/examples/m2m_100?fbclid=IwAR304kICXsffdDMogK4MWf4D7Xeu_3Cbmgu8pBCU_jKcjijCuJfLK7CY9_I
+ Paper: https://arxiv.org/abs/2010.11125

+ This model will not run on 1 V100 GPU, so model parallelism will be needed. 
+ I would expect the state dict to be very similar to mBART, but not sure yet.
+ All I've done is download the state dict, run their command, and asked for help https://github.com/pytorch/fairseq/issues/2772#issuecomment-716152453 when it broke.

Leaving this unassigned in case somebody else wants to take over.

### PR DESCRIPTION
**Pull request by patil-suraj**
# What does this PR do?

Adds the M2M100 model
https://github.com/pytorch/fairseq/tree/master/examples/m2m_100

Fixes #8054

### CODE REVIEW

File: docs/source/model_doc/m2m_100.rst [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+    model = M2M100ForConditionalGeneration.from_pretrained("facebook/m2m100_418M")
+    tokenizer = M2M100Tokenizer.from_pretrained("facebook/m2m100_418M")
+
+    # translate Hindi to French
+    tokenizer.src_lang = "hi"
+    encoded_hi = tokenizer(hi_text, return_tensors="pt")
+    generated_tokens = model.generate(**encoded_hi, forced_bos_token_id=tokenizer.get_lang_id("fr"))
+    tokenizer.batch_decode(generated_tokens, skip_special_tokens=True)
+    # => "La vie est comme une boîte de chocolat."
```
Comment (patil-suraj) [inline docs/source/model_doc/m2m_100.rst:89]: This one got it right  @lhoestq 😉

Comment (lhoestq) [reply docs/source/model_doc/m2m_100.rst:89]: Nice ! 🍫🍫🍫

File: docs/source/model_doc/m2m_100.rst [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+to effectively increase model capacity through a combination of dense scaling and language-specific sparse parameters
+to create high quality models. Our focus on non-English-Centric models brings gains of more than 10 BLEU when directly
+translating between non-English directions while performing competitively to the best single systems of WMT. We
+open-source our scripts so that others may reproduce the data, evaluation, and final M2M-100 model.*
+
+
+Training and Generation
+_______________________________________________________________________________________________________________________
+
+M2M100 is a multilingual encoder-decoder (seq-to-seq) model primarily intended for translation task. As the model is
```
Comment (patrickvonplaten) [inline docs/source/model_doc/m2m_100.rst:41]:
```suggestion
M2M100 is a multilingual encoder-decoder (seq-to-seq) model primarily intended for translation tasks. As the model is
```

File: docs/source/model_doc/m2m_100.rst [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+multilingual it expects the sequences in a different format. A special language id token is used as prefix in both the
```
Comment (patrickvonplaten) [inline docs/source/model_doc/m2m_100.rst:42]:
```suggestion
multilingual it expects the sequences in a certain format. A special language id token is used as prefix in both the
```

File: docs/source/model_doc/m2m_100.rst [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+source and target text. The source text format is is :obj:`[lang_code] X [eos]`, where :obj:`lang_code` is source
```
Comment (patrickvonplaten) [inline docs/source/model_doc/m2m_100.rst:43]:
```suggestion
source and target text. The source text format is :obj:`[lang_code] X [eos]`, where :obj:`lang_code` is source
```

File: docs/source/model_doc/m2m_100.rst [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+language id for source text and target language id for target text, with :obj:`X` being the source or target text
```
Comment (patrickvonplaten) [inline docs/source/model_doc/m2m_100.rst:44]:
```suggestion
language id for both source and target text, with :obj:`X` being the source or target text
```

File: docs/source/model_doc/m2m_100.rst [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+respectively.
```
Comment (patrickvonplaten) [inline docs/source/model_doc/m2m_100.rst:45]:
```suggestion
```

File: docs/source/model_doc/m2m_100.rst [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    tokenizer = M2M100Tokenizer.from_pretrained('facebook/m2m100_418M', src_lang="en", tgt_lang="fr")
+
+    src_text = "Life is like a box of chocolates."
+    tgt_lang = "La vie est comme une boîte de chocolat."
+
+    model_inputs = tokenizer(src_text, return_tensors="pt")
+    with tokenizer.as_target_tokenizer():
+        labels = tokenizer(tgt_text, return_tensors="pt").input_ids
+
+    model(**model_inputs, labels=labels) # forward pass
```
Comment (patrickvonplaten) [inline docs/source/model_doc/m2m_100.rst:63]: (nit) think  `loss = model(**model_inputs, labels=labels).loss` is a bit nicer

File: docs/source/model_doc/m2m_100.rst [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+
+- Generation
+
+    To generate using the M2M100 multilingual translation models, :obj:`eos_token_id` is used as the
```
Comment (patrickvonplaten) [inline docs/source/model_doc/m2m_100.rst:68]:
```suggestion
    M2M100 uses the :obj:`eos_token_id` as the
```

File: docs/source/model_doc/m2m_100.rst [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    :obj:`decoder_start_token_id` and the target language id is forced as the first generated token. To force the
```
Comment (patrickvonplaten) [inline docs/source/model_doc/m2m_100.rst:69]:
```suggestion
    :obj:`decoder_start_token_id` for generation with the target language id being forced as the first generated token. To force the
```

File: src/transformers/models/auto/modeling_auto.py [hunk 10]
```diff
@@ -143,6 +139,10 @@
     LongformerModel,
 )
 from ..lxmert.modeling_lxmert import LxmertForPreTraining, LxmertForQuestionAnswering, LxmertModel
+
+# Add modeling imports here
+# Add modeling imports here
```
Comment (patrickvonplaten) [inline src/transformers/models/auto/modeling_auto.py:144]: this should probably be cleaned up (also cc @LysandreJik this looks like an error in the templates to me)

File: src/transformers/models/m2m_100/configuration_m2m_100.py [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+
+        >>> # Accessing the model configuration
+        >>> configuration = model.config
+    """
+    model_type = "m2m_100"
+    keys_to_ignore_at_inference = ["past_key_values"]
+
+    def __init__(
+        self,
+        vocab_size=128112,
```
Comment (patrickvonplaten) [inline src/transformers/models/m2m_100/configuration_m2m_100.py:102]: that's pretty huge

File: src/transformers/models/m2m_100/modeling_m2m_100.py [hunk 12]
```diff
@@ -0,0 +1,1248 @@
...
+_TOKENIZER_FOR_DOC = "M2M100Tokenizer"
+
+
+M2M_100_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "facebook/m2m100_418M",
+    # See all M2M100 models at https://huggingface.co/models?filter=m2m_100
+]
+
+
+def shift_tokens_right(input_ids: torch.Tensor, pad_token_id: int, decoder_start_token_id: int):
```
Comment (patrickvonplaten) [inline src/transformers/models/m2m_100/modeling_m2m_100.py:58]: could you add a `#Copied from `statement here from Bart maybe?

File: src/transformers/models/m2m_100/modeling_m2m_100.py [hunk 13]
```diff
@@ -0,0 +1,1248 @@
...
+    """
+    Shift input ids one token to the right, and wrap the last non pad token (the <LID> token) Note that MBart does not
```
Comment (patrickvonplaten) [inline src/transformers/models/m2m_100/modeling_m2m_100.py:60]: MBart?

Comment (patrickvonplaten) [reply src/transformers/models/m2m_100/modeling_m2m_100.py:60]: The comment does not seem to be correct

File: src/transformers/models/m2m_100/modeling_m2m_100.py [hunk 14]
```diff
@@ -0,0 +1,1248 @@
...
+
+    index_of_eos = (prev_output_tokens.ne(pad_token_id).sum(dim=1) - 1).unsqueeze(-1)
+    decoder_start_tokens = prev_output_tokens.gather(1, index_of_eos).squeeze()
+    prev_output_tokens[:, 1:] = prev_output_tokens[:, :-1].clone()
+    prev_output_tokens[:, 0] = decoder_start_tokens
+
+    return prev_output_tokens
+
+
+def _make_causal_mask(input_ids_shape: torch.Size, dtype: torch.dtype, past_key_values_length: int = 0):
```
Comment (patrickvonplaten) [inline src/transformers/models/m2m_100/modeling_m2m_100.py:77]: please add `#Copied from` statements

File: src/transformers/models/m2m_100/modeling_m2m_100.py [hunk 15]
```diff
@@ -0,0 +1,1248 @@
...
+    mask_cond = torch.arange(mask.size(-1))
+    mask.masked_fill_(mask_cond < (mask_cond + 1).view(mask.size(-1), 1), 0)
+    mask = mask.to(dtype)
+
+    if past_key_values_length > 0:
+        mask = torch.cat([torch.zeros(tgt_len, past_key_values_length, dtype=dtype), mask], dim=-1)
+    return mask[None, None, :, :].expand(bsz, 1, tgt_len, tgt_len + past_key_values_length)
+
+
+def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int] = None):
```
Comment (patrickvonplaten) [inline src/transformers/models/m2m_100/modeling_m2m_100.py:92]: please add `#Copied from` statements

File: src/transformers/models/m2m_100/modeling_m2m_100.py [hunk 16]
```diff
@@ -0,0 +1,1248 @@
...
+
+    Returns: torch.Tensor
+    """
+    # The series of casts and type-conversions here are carefully balanced to both work with ONNX export and XLA.
+    mask = input_ids.ne(padding_idx).int()
+    incremental_indices = (torch.cumsum(mask, dim=1).type_as(mask) + past_key_values_length) * mask
+    return incremental_indices.long() + padding_idx
+
+
+class M2M100SinusoidalPositionalEmbedding(nn.Module):
```
Comment (patrickvonplaten) [inline src/transformers/models/m2m_100/modeling_m2m_100.py:122]: please add `#Copied from` statements (if possible here not sure with the offset)

File: src/transformers/models/m2m_100/modeling_m2m_100.py [hunk 17]
```diff
@@ -0,0 +1,1248 @@
...
+        input_shape = inputs_embeds.size()[:-1]
+        sequence_length = input_shape[1]
+
+        position_ids = torch.arange(
+            self.padding_idx + 1, sequence_length + self.padding_idx + 1, dtype=torch.long, device=inputs_embeds.device
+        )
+        return position_ids.unsqueeze(0).expand(input_shape).contiguous()
+
+
+class M2M100Attention(nn.Module):
```
Comment (patrickvonplaten) [inline src/transformers/models/m2m_100/modeling_m2m_100.py:194]: please add `#Copied from` statements (if possible here not sure with the offset)

**Review by patrickvonplaten — COMMENTED** [review]
Great work @patil-suraj - could you add as many `#Copied from` statements as possible in the `modeling_...py` file and ping me again for review?

**Comment by patil-suraj** [comment]
Sure, Patrick !

File: docs/source/pretrained_models.rst [hunk 18]
```diff
@@ -365,6 +365,12 @@ For the full list, refer to `https://huggingface.co/models <https://huggingface.
...
 +--------------------+------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------+
+| M2M100             | ``facebook/m2m100_418M``                                   | | 24-layer, 1024-hidden, 16-heads, 418M parameters                                                                                    |
+|                    |                                                            | | multilingual machine translation model for 100 languages                                                                            |
+|                    +------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------+
+|                    | ``facebook/m2m100_1.2B``                                   | | 24-layer, 1024-hidden, 16-heads, 610M parameters                                                                                    |
```
Comment (patil-suraj) [inline docs/source/pretrained_models.rst:371]: still need to port 1.2B checkpoint, will update the layer info after porting

File: tests/test_modeling_m2m_100.py [hunk 19]
```diff
@@ -0,0 +1,372 @@
...
+
+
+@require_torch
+@require_sentencepiece
+@require_tokenizers
+@slow
+class M2M100ModelIntegrationTests(unittest.TestCase):
+    @cached_property
+    def default_tokenizer(self):
+        return M2M100Tokenizer.from_pretrained("valhalla/m2m100_418M")
```
Comment (patil-suraj) [inline tests/test_modeling_m2m_100.py:309]: I Will change the model names, once I push the checkpoints to` facebook/`

File: docs/source/model_doc/m2m_100.rst [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+
+    model = M2M100ForConditionalGeneration.from_pretrained("facebook/m2m100_418M")
+    tokenizer = M2M100Tokenizer.from_pretrained("facebook/m2m100_418M")
+
+    # translate Hindi to French
+    tokenizer.src_lang = "hi"
+    encoded_hi = tokenizer(hi_text, return_tensors="pt")
+    generated_tokens = model.generate(**encoded_hi, forced_bos_token_id=tokenizer.get_lang_id("fr"))
+    tokenizer.batch_decode(generated_tokens, skip_special_tokens=True)
+    # => "La vie est comme une boîte de chocolat."
```
Comment (sgugger) [inline docs/source/model_doc/m2m_100.rst:87]: Since there is a result here, this should use the doctest syntax (so >>> for each line of code, nothing for the return.

File: docs/source/model_doc/m2m_100.rst [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+
+    # translate Chinese to English
+    tokenizer.src_lang = "ar_AR"
+    encoded_zh = tokenizer(chinese_text, return_tensors="pt")
+    generated_tokens = model.generate(**encoded_zh, forced_bos_token_id=tokenizer.get_lang_id("en"))
+    tokenizer.batch_decode(generated_tokens, skip_special_tokens=True)
+    # => "Life is like a box of chocolate."
```
Comment (sgugger) [inline docs/source/model_doc/m2m_100.rst:94]: Same here.

File: src/transformers/models/auto/modeling_auto.py [hunk 22]
```diff
@@ -21,8 +21,6 @@
 from ...configuration_utils import PretrainedConfig
 from ...file_utils import add_start_docstrings
 from ...utils import logging
-
-# Add modeling imports here
```
Comment (sgugger) [inline src/transformers/models/auto/modeling_auto.py:25]: This should not be removed in this PR as it is used by the model templates to detect where the new model imports should be added.

File: src/transformers/models/auto/modeling_auto.py [hunk 23]
```diff
@@ -66,8 +64,6 @@
     CamembertForTokenClassification,
     CamembertModel,
 )
-
-# Add modeling imports here
```
Comment (sgugger) [inline src/transformers/models/auto/modeling_auto.py:70]: This one can stay as it was added by mistake.

File: src/transformers/models/m2m_100/configuration_m2m_100.py [hunk 24]
```diff
@@ -0,0 +1,164 @@
...
+
+    Configuration objects inherit from :class:`~transformers.PretrainedConfig` and can be used to control the model
+    outputs. Read the documentation from :class:`~transformers.PretrainedConfig` for more information.
+
+
+    Args:
+        vocab_size (:obj:`int`, `optional`, defaults to 50265):
+            Vocabulary size of the M2M100 model. Defines the number of different tokens that can be represented by the
+            :obj:`inputs_ids` passed when calling :class:`~transformers.M2M100Model` or
+            :class:`~transformers.TFM2M100Model`.
```
Comment (sgugger) [inline src/transformers/models/m2m_100/configuration_m2m_100.py:44]: No TF model
```suggestion
            :obj:`inputs_ids` passed when calling :class:`~transformers.M2M100Model`.
```

File: src/transformers/models/m2m_100/modeling_m2m_100.py [hunk 25]
```diff
@@ -0,0 +1,1312 @@
...
+
+    @property
+    def dummy_inputs(self):
+        pad_token = self.config.pad_token_id
+        input_ids = torch.tensor([[0, 6, 10, 4, 2], [0, 8, 12, 2, pad_token]], device=self.device)
+        dummy_inputs = {
+            "attention_mask": input_ids.ne(pad_token),
+            "input_ids": input_ids,
+        }
+        return dummy_inputs
```
Comment (sgugger) [inline src/transformers/models/m2m_100/modeling_m2m_100.py:543]: Why is this here? PyTorch models don't require dummy inputs.

File: src/transformers/models/m2m_100/tokenization_m2m_100.py [hunk 26]
```diff
@@ -0,0 +1,327 @@
...
+        unk_token (:obj:`str`, `optional`, defaults to :obj:`"<unk>"`):
+            The unknown token. A token that is not in the vocabulary cannot be converted to an ID and is set to be this
+            token instead.
+        pad_token (:obj:`str`, `optional`, defaults to :obj:`"<pad>"`):
+            The token used for padding, for example when batching sequences of different lengths.
+
+    Examples::
+
+        >>> from transformers import M2M100Tokenizer
+        >>> tokenizer = M2M100Tokenizer.from_pretrained("facebook/m2m100_418M, src_lang="en_XX", tgt_lang="ro_RO")
```
Comment (sgugger) [inline src/transformers/models/m2m_100/tokenization_m2m_100.py:83]: Love the set at init! But it seems from what's below that the language codes are not the same as mBART.
```suggestion
        >>> tokenizer = M2M100Tokenizer.from_pretrained("facebook/m2m100_418M, src_lang="en", tgt_lang="ro")
```

**Review by sgugger — APPROVED** [review]
Great addition! Out of curiosity, what is missing to get a fast tokenizer like for mBART?

File: src/transformers/models/m2m_100/configuration_m2m_100.py [hunk 27]
```diff
@@ -0,0 +1,164 @@
...
+        >>> from transformers import M2M100Model, M2M100Config
+
+        >>> # Initializing a M2M100 m2m_100_418M style configuration
+        >>> configuration = M2M100Config()
+
+        >>> # Initializing a model from the m2m_100_418M style configuration
+        >>> model = M2M100Model(configuration)
+
+        >>> # Accessing the model configuration
+        >>> configuration = model.config
```
Comment (LysandreJik) [inline src/transformers/models/m2m_100/configuration_m2m_100.py:95]: This needs to be indented otherwise it doesn't render well in the docs

**Review by LysandreJik — APPROVED** [review]
This looks great! Would it be hard to implement the fast tokenizer as well?

File: src/transformers/models/m2m_100/modeling_m2m_100.py (hunk 25 above)
Comment (patrickvonplaten) [reply src/transformers/models/m2m_100/modeling_m2m_100.py:543]: Actually, I don't know why there were there in the first place in Bart & I just left them. But I think we should do a clean-up where we remove all those `dummy_inputs` from all Bart models. I'm ok for leaving it in this PR for now and remove all of them in a follow-up PR.

File: src/transformers/models/m2m_100/configuration_m2m_100.py [hunk 28]
```diff
@@ -0,0 +1,164 @@
...
+        is_encoder_decoder=True,
+        activation_function="relu",
+        d_model=1024,
+        dropout=0.1,
+        attention_dropout=0.1,
+        activation_dropout=0.0,
+        init_std=0.02,
+        decoder_start_token_id=2,
+        scale_embedding=True,
+        gradient_checkpointing=False,
```
Comment (patrickvonplaten) [inline src/transformers/models/m2m_100/configuration_m2m_100.py:122]: Docs for gradient_checkpointing are missing (are they missing in the templates as well?) -> could you maybe add them there in this case as well @patil-suraj ?

File: src/transformers/models/m2m_100/modeling_m2m_100.py [hunk 29]
```diff
@@ -0,0 +1,1312 @@
...
+    return inverted_mask.masked_fill(inverted_mask.bool(), torch.finfo(dtype).min)
+
+
+def create_position_ids_from_input_ids(input_ids, padding_idx, past_key_values_length=0):
+    """
+    Replace non-padding symbols with their position numbers. Position numbers begin at padding_idx+1. Padding symbols
+    are ignored. This is modified from fairseq's `utils.make_positions`.
+
+    Args:
+        x: torch.Tensor x:
```
Comment (patrickvonplaten) [inline src/transformers/models/m2m_100/modeling_m2m_100.py:111]: can you improve the docstring here? there is not input `x`

File: src/transformers/models/m2m_100/modeling_m2m_100.py [hunk 30]
```diff
@@ -0,0 +1,1312 @@
...
+    # The series of casts and type-conversions here are carefully balanced to both work with ONNX export and XLA.
+    mask = input_ids.ne(padding_idx).int()
+    incremental_indices = (torch.cumsum(mask, dim=1).type_as(mask) + past_key_values_length) * mask
+    return incremental_indices.long() + padding_idx
+
+
+class M2M100SinusoidalPositionalEmbedding(nn.Module):
+    """This module produces sinusoidal positional embeddings of any length."""
+
+    def __init__(self, num_positions: int, embedding_dim: int, padding_idx: Optional[int] = None):
```
Comment (patrickvonplaten) [inline src/transformers/models/m2m_100/modeling_m2m_100.py:124]: let's remove the padding index -> there should be no padding index in the positional embeddings. See [this](https://github.com/huggingface/transformers/pull/10200) discussion.

File: src/transformers/models/m2m_100/modeling_m2m_100.py [hunk 31]
```diff
@@ -0,0 +1,1312 @@
...
+        super().__init__()
+        self.offset = 2
+        self.embedding_dim = embedding_dim
+        self.padding_idx = padding_idx
+        self.weights = self.get_embedding(num_positions + self.offset, embedding_dim, padding_idx)
+        self.register_buffer("_float_tensor", torch.FloatTensor(1))
+
+    @staticmethod
+    def get_embedding(num_embeddings: int, embedding_dim: int, padding_idx: Optional[int] = None):
```
Comment (patrickvonplaten) [inline src/transformers/models/m2m_100/modeling_m2m_100.py:133]: also let's remove the padding_idx logic here -> there shouldn't be one for positions

File: src/transformers/models/m2m_100/modeling_m2m_100.py [hunk 32]
```diff
@@ -0,0 +1,1312 @@
...
+        >>> from transformers import M2M100Tokenizer, M2M100ForConditionalGeneration
+
+        >>> model = M2M100ForConditionalGeneration.from_pretrained('facebook/m2m100_418M')
+        >>> tokenizer = M2M100Tokenizer.from_pretrained('facebook/m2m100_418M')
+
+        >>> text_to_translate = "Life is like a box of chocolates"
+        >>> model_inputs = tokenizer(text_to_translate, return_tensors='pt')
+
+        >>> # translate to French
+        >>> gen_tokens = model.generate( **model_inputs, forced_bos_token_id=tok.get_lang_id("fr"))
```
Comment (patrickvonplaten) [inline src/transformers/models/m2m_100/modeling_m2m_100.py:574]: really like the new "forced_bos_token_id" design - nice!

File: src/transformers/models/m2m_100/modeling_m2m_100.py [hunk 33]
```diff
@@ -0,0 +1,1312 @@
...
+
+        if embed_tokens is not None:
+            self.embed_tokens = embed_tokens
+        else:
+            self.embed_tokens = nn.Embedding(config.vocab_size, embed_dim, self.padding_idx)
+
+        self.embed_positions = M2M100SinusoidalPositionalEmbedding(
+            config.max_position_embeddings,
+            embed_dim,
+            self.padding_idx,
```
Comment (patrickvonplaten) [inline src/transformers/models/m2m_100/modeling_m2m_100.py:672]: think we sholdu remove the padding idx

**Review by patrickvonplaten — APPROVED** [review]
Great job! Looks very clean & thanks for adding all those copied from statements!

Left a couple of smaller improvement suggestions:

1) One doc was missing
2) Think we should not continue this practice of giving the padding_idx a special behavior in the postiional embeddings
3) We should move all checkpoints to `facebook` before merging

**Comment by patil-suraj** [comment]
I’ve addressed all the review comments, and all the slow/fast tests are now passing.

I didn’t add fast tokenizer because `M2M100` is `sentencepiece`  based tokenizer, but it uses `sentencepiece`  for just tokenizing and then uses a vocab file to convert the tokens to ids and ids to tokens. So our current `SpmConverter` doesn’t work for this. I’ll try to add fast tokenizer in a follow-up PR.

Merging!

**Comment by Muennighoff** [comment]
> I’ve addressed all the review comments, and all the slow/fast tests are now passing.
> 
> I didn’t add fast tokenizer because `M2M100` is `sentencepiece` based tokenizer, but it uses `sentencepiece` for just tokenizing and then uses a vocab file to convert the tokens to ids and ids to tokens. So our current `SpmConverter` doesn’t work for this. I’ll try to add fast tokenizer in a follow-up PR.
> 
> Merging!

Hey, I was wondering if there's any progress on a Fast Tokenizer for M2M or if any help can be needed?
Thanks :)
